test/services/database/comment_repository_test.dart

```diff
@@ -0,0 +1,249 @@
+import "package:proxima/models/database/post/post_firestore.dart";
+import "package:proxima/models/database/post/post_id_firestore.dart";
+import "package:proxima/services/database/comment_repository_service.dart";
+import "package:proxima/services/database/firestore_service.dart";
+import "package:proxima/services/database/post_repository_service.dart";
+
+import "../../mocks/data/firestore_comment.dart";
+import "../../mocks/data/geopoint.dart";
+import "../../mocks/data/post_data.dart";
+
+void main() {
+  group("Testing comment repository", () {
+    late FakeFirebaseFirestore fakeFirestore;
+    late PostRepositoryService postRepository;
+    late CommentRepositoryService commentRepository;
+
+    late PostIdFirestore postId;
+    late CollectionReference<Map<String, dynamic>> commentsSubCollection;
+    late DocumentReference<Map<String, dynamic>> postDocument;
+    late CommentFirestoreGenerator commentGenerator;
+
+    Future<void> addComment(CommentFirestore comment) async {
+      await commentsSubCollection
+          .doc(comment.id.value)
+          .set(comment.data.toDbData());
+
+      await postDocument.update({
+        PostData.commentCountField: FieldValue.increment(1),
+      });
+    }
+
+    Future<List<CommentFirestore>> addComments(int number) async {
+      final comments = <CommentFirestore>[];
+
+      for (var i = 0; i < number; i++) {
+        final comment = commentGenerator.createRandomComment();
+        await addComment(comment);
+        comments.add(comment);
+      }
+
+      return comments;
+    }
+
+    /// The setup add a single post with no comments
+    setUp(() async {
+      fakeFirestore = FakeFirebaseFirestore();
+      postRepository = PostRepositoryService(
+        firestore: fakeFirestore,
+      );
+
+      // We get the comment repository from the provider container
+      // because it allows to check that the provider constructs
+      // the repository correctly
+      final container = ProviderContainer(
+        overrides: [
+          firestoreProvider.overrideWithValue(fakeFirestore),
+          postRepositoryProvider.overrideWithValue(postRepository),
+        ],
+      );
+
+      commentRepository = container.read(commentRepositoryProvider);
+
+      final postData = PostDataGenerator().postData;
+      postId = await postRepository.addPost(postData, userPosition0);
+
+      postDocument = fakeFirestore
+          .collection(
+            PostFirestore.collectionName,
+          )
+          .doc(postId.value);
+
+      commentsSubCollection = postDocument.collection(
+        CommentFirestore.subCollectionName,
+      );
+
+      commentGenerator = CommentFirestoreGenerator();
+    });
+
+    group("getting comments", () {
+      test("should return an empty list if there are no comments", () async {
+        final comments = await commentRepository.getComments(postId);
+
+        expect(comments, isEmpty);
+      });
+
+      test(
```

Comment:
I am not really a fan of that kind of testing where the test basically duplicate the behavior of the supposed component under test, and then we just check that the two codes are actually doing the same thing.
What I prefer is to simply use the component under test and then assert if the resulting behavior is the one we expected (black box testing).
I don't know what @JoachimFavre (other reviewer) thinks about this topic ?

Reply:
I am not sure I understand really well.
Because from my point of view I am not comparing that my test code is doing the same as my production code.

I have added some helper functions to facilitate the creation of the conditions for the tests but that's it.
Then I check that I get the expected behavior depending on the condition.

For instance here I check that I indeed get the comments that are on the database:
```dart
      test(
          "should return the comments of a post when there are multiple comments",
          () async {
        final comments = await addComments(5);

        final actualComments = await commentRepository.getComments(postId);

        expect(actualComments, unorderedEquals(comments));
      });
```

Here, `addComments` is only used to put the initial comments on the database. I don't really see how I could do it another way ?

Reply:
Yes but that's exactly my point...
Your testing "helper" method replicates the actual implementation and introduces a lot of coupling between the two.
I would use the add comment method from the repository directly and check for correct outside observable behavior instead of comparing it with a "duplicate" method.

Reply:
Ohh ok I see.
I think the problem with this approach is that the tests are no longer unitary for the `CommentRepository`.

For instance, the above test would rely on `CommentRepository.addComment` and `CommentRepository.getComments`.

This can be an issue because for instance if `CommentRepository.addComment` would simply store the comments in the memory instead of putting them on firestore, we would have no way of checking that the `getComments` doesn't actually give the comments that are on the database.
That's one of the main reasons for why I chose to use helper functions to ensure that the data are actually coming from the database.

Reply:
It is still unit testing, go with the mantra "Just don't worry about it and test the exposed behaviour". Check out https://stackoverflow.com/questions/5958895/how-can-i-correctly-unit-test-methods-that-depend-on-each-other it's a great thread.

Reply:
Yes exactly :) Thank you very much.